### PR TITLE
Manifold Mixup: feature-level interpolation for OOD generalization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model overfits to the discrete distribution of training samples. At convergence, encoded feature representations cluster tightly around training examples, leaving gaps in the feature space that OOD samples fall into. **Manifold Mixup** (Verma et al., ICML 2019) addresses this by creating virtual training examples through interpolation of encoded features and targets.

During training, for each batch:
1. Sample λ ~ Beta(α, α) where α controls interpolation strength
2. Create shuffled batch indices
3. After the backbone encoding (preprocess MLP + TransolverBlocks), interpolate encoded features: `fx_mix = λ * fx + (1-λ) * fx[perm]`
4. Interpolate targets: `y_mix = λ * y + (1-λ) * y[perm]`
5. Compute loss on mixed features/targets

This fills the feature space between training samples with plausible interpolations, creating a smoother loss landscape and better OOD generalization.

**Why it might work here:**
- The biggest performance gaps are on OOD metrics (p_oodc 7.64 vs ensemble 6.6, p_re 6.30 vs ensemble 5.8)
- Mixup is proven to improve OOD generalization in vision (ImageNet), NLP, and scientific ML
- Feature-level Mixup (manifold) is more effective than input-level Mixup because the encoded space is more linearly separable
- With α=0.2 (mild interpolation), most λ values are near 0 or 1, so the model mostly sees real samples with occasional mild interpolation

**Key distinction from input noise augmentation (#2235, alphonse):**
- Input noise: adds random perturbation to each sample independently
- Manifold Mixup: creates convex combinations of REAL encoded representations → explores the space BETWEEN real samples

**Confidence:** Medium-high. Well-established technique with strong theoretical backing (Verma et al., 2019; Mixup: Zhang et al., ICLR 2018). Main risk: pressure fields may not interpolate linearly in encoded space (physics is nonlinear). Using a small α (0.2) mitigates this.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
manifold_mixup: bool = False     # feature-level mixup for OOD generalization
mixup_alpha: float = 0.2         # Beta distribution parameter (small = mild mixing)
```

### Step 2: Implement Mixup in the training loop

In the **training loop only**, AFTER the backbone forward pass (all TransolverBlocks) but BEFORE the output heads (pressure decoder, velocity decoder, SRF heads):

```python
if cfg.manifold_mixup and model.training:
    # Sample mixing coefficient
    lam = torch.distributions.Beta(cfg.mixup_alpha, cfg.mixup_alpha).sample().to(device)
    
    # Random permutation for pairing
    perm = torch.randperm(fx.shape[0], device=device)
    
    # Mix encoded features
    fx = lam * fx + (1.0 - lam) * fx[perm]
    
    # Mix ALL targets correspondingly
    # Find where the target tensors are (pressure, velocity) and mix them:
    target_p = lam * target_p + (1.0 - lam) * target_p[perm]
    target_ux = lam * target_ux + (1.0 - lam) * target_ux[perm]
    target_uy = lam * target_uy + (1.0 - lam) * target_uy[perm]
    
    # Also mix any masks/weights that are per-sample
    # e.g., surface masks, boundary IDs for loss computation
    # IMPORTANT: binary masks should NOT be mixed — use the original sample's masks
    # Only continuous targets should be interpolated
```

**CRITICAL:** 
- Only mix in the training loop. Validation must use unmixed data.
- Mix features AND targets with the SAME λ and permutation.
- Do NOT mix binary masks (surface_mask, boundary_id). Use the original sample's masks for loss computation.
- Do NOT mix when batch_size=1 (need at least 2 samples to mix).

### Step 3: Handle SRF heads

The SRF heads (SurfaceRefinementHead, AftFoilRefinementHead) operate on subsets of nodes. After mixing encoded features `fx`, the SRF heads should process the mixed features normally. The mixed targets should be used for the SRF loss computation.

### Step 4: Handle PCGrad compatibility

PCGrad separates gradients by loss component. Mixup doesn't change which loss components exist — it only changes the features and targets. PCGrad should work normally with mixed inputs.

### Step 5: torch.compile compatibility

All operations (Beta sampling, indexing, linear interpolation) are compile-safe. The `torch.distributions.Beta` call might need to be outside the compiled region — sample λ before entering the compiled forward pass.

### Step 6: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent nezuko --wandb_name "nezuko/manifold-mixup-s42" \
  --wandb_group "round18/manifold-mixup" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --manifold_mixup --mixup_alpha 0.2

# Seed 73 — identical but --seed 73 --wandb_name "nezuko/manifold-mixup-s73"
```

### Step 7: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent nezuko --wandb_name "nezuko/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results

### Setup
- **Hypothesis**: Manifold Mixup at the hidden feature level (after backbone TransolverBlocks, before the last output block) forces the model to interpolate between real encoded representations, creating smoother feature-space loss landscape and improving OOD generalization.
- **Implementation**: Sample λ ~ Beta(0.2, 0.2), mix hidden state fx after blocks[:-1], also mix normalized targets y_norm with same λ and permutation. Passed via model input dict `{"manifold_mix_lam": lam, "manifold_mix_perm": perm}`. Mixed sample_stds used for pred denormalization.
- **Runs**: 2 seeds (42 and 73), WANDB_MODE=offline

### Metrics

| Split | Seed 42 (5fiy4jn6) | Seed 73 (bsil84ca) | 2-seed avg | Baseline (PR #2213) | Δ |
|---|---|---|---|---|---|
| p_in | 13.954 | 13.818 | **13.886** | 11.979 | **+15.9%** ✗ |
| p_tan | 34.9 | 35.3 | **35.10** | 28.341 | **+23.9%** ✗ |
| p_oodc | 9.9 | 9.6 | **9.75** | 7.643 | **+27.6%** ✗ |
| p_re | 7.4 | 7.5 | **7.45** | 6.300 | **+18.3%** ✗ |

Best epochs: s42=145, s73=145.

### Analysis

Clear regression across all metrics. Manifold Mixup at the hidden state is significantly hurting the model.

**Why it failed here:** The key difference between CFD mesh data and image/NLP data where Manifold Mixup works is **non-uniform geometric node distributions**. In image classification, every sample has the same spatial grid and pixels at position (i,j) across samples represent the same concept. Here, every mesh has a different set of 80K-160K nodes at different spatial locations. When we mix  and , node  in sample  may be near the leading edge, while node  in sample  is deep in the wake — there's no semantic alignment between node indices across samples. The mixed hidden features for node  don't correspond to any physically meaningful representation, and the model can't learn to produce correct predictions at the original node positions from this incoherent mixed representation.

Additionally, the per-sample std normalization (which normalizes the scale of the training targets) creates a further inconsistency: the mixed hidden features come from two samples with potentially very different geometric scales (single-foil vs tandem, different Re), and the convex combination of their normalized targets doesn't match the prediction scale of the mixed hidden features.

### Suggested follow-ups
1. **Node-aligned Manifold Mixup**: Mix only samples that have the same mesh (single-foil with single-foil, tandem with tandem), or only single-foil samples with themselves. This avoids node index misalignment across configurations.
2. **Feature-level dropout regularization**: Instead of mixing, apply structured dropout on the hidden features (e.g., drop entire attention head outputs stochastically) to regularize.
3. **Consistent-node mixup via spatial matching**: For each pair of samples, find nodes with similar spatial coordinates (kNN matching) and mix at aligned nodes — similar to spatial interpolation. More complex but geometrically meaningful.
4. **Standard input Mixup with geometry-preserving coordinates**: Mix only the physical fields (y targets) while keeping node positions fixed from one sample — this is a simpler form of supervised noise that doesn't require hidden state mixing.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2237 because ADVISOR: Closing — all metrics regressed substantially (p_in +15.9%, p_oodc +27.6%, p_tan +23.9%, p_re +18.3%). Your analysis is excellent — the core issue is that mesh node indices have NO spatial alignment across samples. In images, pixel (i,j) means the same thing across samples. Here, node k in sample A may be at the leading edge while node k in sample B is in the wake. Mixing hidden states at unaligned nodes produces physically incoherent representations. The per-sample std normalization further compounds this. Manifold mixup is fundamentally inapplicable to variable-mesh point cloud architectures. Thank you for the thorough write-up..
